kong/tools/queue.lua

```diff
@@ -0,0 +1,400 @@
+    queue[option] = value
+  end
+
+  queue = setmetatable(queue, Queue_mt)
+
+  kong.timer:named_at("queue " .. name, 0, function(_, q)
+    while q:count() > 0 do
+      q:log_debug("processing queue")
+      q:process_once()
+    end
+    q:log_debug("done processing queue")
+    queues[name] = nil
+  end, queue)
+
+  queues[name] = queue
+
+  queue:log_debug("queue created")
+
+  return queue
+end
+
+
+-------------------------------------------------------------------------------
+-- Log a message that includes the name of the queue for identification purposes
+-- @param self Queue
+-- @param level: log level
+-- @param formatstring: format string, will get the queue name and ": " prepended
+-- @param ...: formatter arguments
+function Queue:log(handler, formatstring, ...)
+  local message = string.format("queue %s: %s", self.name, formatstring)
+  if select('#', ...) > 0 then
+    return handler(string.format(message, unpack({...})))
+  else
+    return handler(message)
+  end
+end
+
+function Queue:log_debug(...) self:log(kong.log.debug, ...) end
+function Queue:log_info(...) self:log(kong.log.info, ...) end
+function Queue:log_warn(...) self:log(kong.log.warn, ...) end
+function Queue:log_err(...) self:log(kong.log.err, ...) end
+
+
+function Queue:count()
+  return self.back - self.front
+end
+
+
+-- Delete the frontmost entry from the queue and adjust the current utilization variables.
+function Queue:delete_frontmost_entry()
+  if self.max_bytes then
+    -- If max_bytes is set, reduce the currently queued byte count by the
+    self.bytes_queued = self.bytes_queued - #self.entries[self.front]
+  end
+  self.entries[self.front] = nil
+  self.front = self.front + 1
+  if self.front == self.back then
+    self.front = 1
+    self.back = 1
+  end
+end
+
+
+-- Drop the oldest entry, adjusting the semaphore value in the process.  This is
+-- called when the queue runs out of space and needs to make space.
+function Queue:drop_oldest_entry()
+  assert(self.semaphore:count() > 0)
+  self.semaphore:wait(0)
+  self:delete_frontmost_entry()
+end
+
+
+-- Process one batch of entries from the queue.  Returns truthy if entries were processed, falsy if there was an
+-- error or no items were on the queue to be processed.
+function Queue:process_once()
+  local ok, err = self.semaphore:wait(0)
+  if not ok then
+    if err ~= "timeout" then
+      -- We can't do anything meaningful to recover here, so we just log the error.
+      self:log_err('error waiting for semaphore: %s', err)
+    end
+    return
+  end
+  local data_started = now()
+
+  local entry_count = 1
+
+  -- We've got our first entry from the queue.  Collect more entries until max_coalescing_delay expires or we've collected
+  -- max_batch_size entries to send
+  while entry_count < self.max_batch_size and (now() - data_started) < self.max_coalescing_delay and not ngx.worker.exiting() do
+    ok, err = self.semaphore:wait(((data_started + self.max_coalescing_delay) - now()) / 1000)
```

Comment:
@hanshuebner 

I've tried the latest version of the queue and I came to note that the queue is processed almost immediately after kong proxied a request, when using `max_coalescing_delay=20`. This unit is supposed to be seconds but I believe it is used as ms here. 
After setting the `max_coalescing_delay=20000` the queue is processed after 20s. 

When testing the shutdown behavior in that case Kong forcefully stops now. This wasn't the case previously. 
At my last review the queue had a named_every timer that handled the shutdown with flushing the queue.
https://github.com/Kong/kong/blob/2f188be0c82c89038624fc3c75c68181897f5568/kong/tools/queue.lua#L157

This seems to be replaced with a single timer to decouple the queue process 
https://github.com/Kong/kong/blob/8189ce33978dd3e6cc295d618642ee6d70224f71/kong/tools/queue.lua#L135
However I don't see where the premature event is handled. The queue is stuck here waiting for the semaphore.

I'm not sure if this is the right place to post this since the PR is already merged.

Reply:
@27ascii Thank you for testing things out!  Now that the PR is merged, new issues would be preferred for problem reports going forward.  I'll create internal tickets for the two problems that you've reported above, though, so no need to open issues unless you feel like it.

> After setting the max_coalescing_delay=20000 the queue is processed after 20s.

This is a bug that we're going to fix soon.

> When testing the shutdown behavior in that case Kong forcefully stops now. [...]
> I don't see where the premature event is handled.  The queue is stuck here waiting for the semaphore.

Good catch!  I need to dig a little deeper here, but right now, the situation is like you describe:  If the shutdown grace period is shorter than `max_coalescing_delay`, data is likely to be lost while waiting for the semaphore.  Expect a fix for this as well.

Thank you again!
Hans

Reply:
> I don't see where the premature event is handled.

Just as a side note regarding this:  We're now using an `at` instead of an `every` timer, which means that there is no need to distinguish whether the handler was invoked normally or in the unlikely edge case that is invoked between when it was scheduled to run (immediately) and a shutdown event arriving before it is invoked.  In both cases, the queue needs to be flushed and that happens without further timer involvement.  This is also the reason why graceful shutdown now works to some extent - The previous implementation needed to schedule new timers while flushing, which is not possible during shutdown.